### PR TITLE
Dynamic inventory

### LINKED ISSUE #7: Allow users to specify their own custom domains
...in addition to `*.hgv.dev`, of course.

This is going to be a big feature request, I think, so we should be out ahead of it.

### LINKED ISSUE #15: Use dynamic inventory to allow custom host names
I'm liking the notion of making hhvm.\*, php.\*, cache.hhvm.\*, cache.php.\*, etc. all work on single instances of WordPress, as per @markkelnar's ideas in HipChat. I think it would be extra-compelling if we could allow for dynamic inventory creation and allow folks to add their own lists of host names/sites to a single file (likely a YAML one) stored somewhere outside of the provisioning tree. Then we can just run the `-i` on Ansible and point it to a shell, PHP, whatever script that populates the JSON catalog with our defaults *and* reads the YAML file to pull in the custom requests. 

We could also do this for per-site plugin and theme loadouts.

Downside: Regular people would have to learn YAML.
Upside: Much more flexibility.

We would also need to dynamically populate the host list in the `Vagrantfile` using some Ruby in order to make sure the new domains get picked up by the `vagrant-hostsupdater` plugin.

This solution should ideally solve

* [ ] #7 
* [ ] #16

### LINKED ISSUE #44: It's possible to have one Hgv instance and multiple sites?
I was wondering if there is a quick way of adding a new subdomain, e.g. `foo.hgv.dev` and creating a new DB for that subdomain as well (with or without installing WP, this is less important)

Thanks!

### PR DESCRIPTION
**Pull request by zamoose**
This creates a new role, wpsites, that allows users to create custom WordPress sites and have them automatically provisioned.

* [ ] @zamoose
* [ ] @markkelnar 

Fixes #7. Fixes #15. Fixes #44.

### CODE REVIEW

**Comment by michaelbeil** [comment]
:cake: :cake: :cake:

File: provisioning/roles/wpsites/tasks/db.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+---
+- name: "Provision WP Databases"
+  mysql_db: name="wpe_{{ item.site | regex_replace('\.','') }}" state=present
+  with_items: wpsites
+  tags: [ 'wpsites' ]
+
+- name: "Grant WP users access to WP DBs"
+  mysql_user: name="wpe_{{ item.site | regex_replace('\.','') }}" priv="wpe_{{ item.site | regex_replace('\.','') }}.*:ALL" host="%" password=wordpress state=present
```
Comment (markkelnar) [inline provisioning/roles/wpsites/tasks/db.yml]: Gonna harass @zamoose about not using multi-line tasks here.

File: provisioning/roles/wpsites/tasks/listeners.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  template:
+    src: etc/nginx/conf.d/wpcache.conf
+    dest: /etc/nginx/conf.d/cache.{{ item.site }}.conf
+    owner: root
+    group: root
+    mode: 0644
+  with_items: wpsites
+  tags: [ 'wpsites' ]
+
+- name: "Enable PML viewing for PHP virtualhosts"
```
Comment (markkelnar) [inline provisioning/roles/wpsites/tasks/listeners.yml]: PML for PHP backend, but not for HHVM backend?

File: provisioning/roles/wpsites/tasks/db.yml (hunk 1 above)
Comment (zamoose) [inline provisioning/roles/wpsites/tasks/db.yml]: Fixed in 9400556703ca935667d7a124461052ffee19ca32.

File: provisioning/roles/wpsites/templates/etc/nginx/conf.d/wpsites.conf [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	# Pass all .php files onto a php-fpm/php-fcgi server.
+	location ~ \.php$ {
+		# Enable for xhprof tracing
+		fastcgi_param PHP_VALUE "auto_prepend_file={{ xhprof_root }}/external/header.php \n auto_append_file={{ xhprof_root }}/external/footer.php";
+
+		fastcgi_split_path_info ^(.+\.php)(/.+)$;
+		fastcgi_pass {{ wpbackend }};
+		fastcgi_intercept_errors on;
+		fastcgi_index index.php;
+		fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
```
Comment (markkelnar) [inline provisioning/roles/wpsites/templates/etc/nginx/conf.d/wpsites.conf]: Should probably add the fastcgi_buffers and _size to this Nginx config file as added in https://github.com/wpengine/hgv/pull/127.

Comment (zamoose) [inline provisioning/roles/wpsites/templates/etc/nginx/conf.d/wpsites.conf]: Added in 5774892.

File: Vagrantfile [hunk 4]
```diff
@@ -7,6 +7,37 @@ VAGRANTFILE_API_VERSION = "2"
...
+        tmpdomains.push('hhvm.' << domain)
+    end
+    customdomains += tmpdomains
+
+    domains_array = domains_array | customdomains
+
+    customdomains.each do |domain|
+        domains_array.push('cache.' << domain)
+    end
+end
```
Comment (markkelnar) [inline Vagrantfile]: Looks good on this customdomains piece in the Vagrantfile.

File: provisioning/roles/wpsites/tasks/listeners.yml (hunk 2 above)
Comment (zamoose) [inline provisioning/roles/wpsites/tasks/listeners.yml]: Neither are getting picked up in the PML web UI on my testbed. Scratching my head right now.

**Comment by markkelnar** [comment]
Why bring the listener and wordpress roles back together?  Doesn't look like much functional differences.  Why not change them to support the array format you've created and just pass those items in.  Seems less traumatic to the codebase and less duplication of code we'll have to worry about maintaining.

File: provisioning/roles/wpsites/tasks/listeners.yml (hunk 2 above)
Comment (zamoose) [inline provisioning/roles/wpsites/tasks/listeners.yml]: Added in 288885e.

File: provisioning/roles/common/tasks/swap.yml [hunk 5]
```diff
@@ -1,10 +1,11 @@
 ---
 - name: Create swap file
   command: dd if=/dev/zero of=/swapfile bs=1024 count=2048k creates=/swapfile
+  register: write_swapfile
 
 - name: Setup swap area
   command: mkswap /swapfile
-  when: ansible_swaptotal_mb < 1
+  when: ansible_swaptotal_mb < 1 and write_swapfile.changed
```
Comment (markkelnar) [inline provisioning/roles/common/tasks/swap.yml]: This swap stuff doesn't appear to be related to the custom domains change.  It should really be in it's own feature branch and pull request.

File: provisioning/roles/sites/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,20 @@
+---
+- name: "Set up Non-WP docroots"
```
Comment (markkelnar) [inline provisioning/roles/sites/tasks/main.yml]: Can you help me understand the non-WP need again?

File: provisioning/roles/wpsites/templates/etc/nginx/conf.d/wpcache.conf [hunk 7]
```diff
@@ -0,0 +1,27 @@
+server {
+        listen          80;
+        server_name     cache.{{ item.site }};
```
Comment (markkelnar) [inline provisioning/roles/wpsites/templates/etc/nginx/conf.d/wpcache.conf]: How is this item.site associated/mapped to each backend (hhvm, php)?

File: provisioning/roles/wpsites/templates/etc/nginx/conf.d/wpsites.conf [hunk 8]
```diff
@@ -0,0 +1,50 @@
+server {
+	listen          80;
+	{% if wpbackend == "hhvm" %}
+	server_name     hhvm.{{ item.site }};
```
Comment (markkelnar) [inline provisioning/roles/wpsites/templates/etc/nginx/conf.d/wpsites.conf]: Does this mean that each domain or custom domain provided by the customer will be prepended with 'hhvm.'.  And non-hhvm will default to the domain the user specified?  If we went with some sort of logic like this, I'd kind of prefer it the other way.  That the default behavior is HHVM as the PHP processor and the other is prepended with a PHP.  

Let's define the behavior we want before making any further changes here.

File: provisioning/roles/common/tasks/swap.yml (hunk 5 above)
Comment (zamoose) [inline provisioning/roles/common/tasks/swap.yml]: Agreed. I had to make an emergency scramble to get it provisioning. I can break it out.

File: provisioning/roles/wpsites/templates/etc/nginx/conf.d/wpsites.conf (hunk 8 above)
Comment (zamoose) [inline provisioning/roles/wpsites/templates/etc/nginx/conf.d/wpsites.conf]: Yes, that's correct. I was considering this to be a situation where people were emulating their current site and then taking HHVM as the edge case -- use the regular one to be the "control" in the HHVM experiment.

File: provisioning/roles/sites/tasks/main.yml (hunk 6 above)
Comment (zamoose) [inline provisioning/roles/sites/tasks/main.yml]: Two use cases:

1. Breaking out core non-WP sites (e.g. admin) into a role that isn't as hard-coded.
2. Making HGV an attractive development platform for people looking to integrate with WordPress.

With the emergence of the WP-API, WordPress is increasingly going to be used as a cog in larger systems capable of consuming and emitting JSON. Being able to develop, for instance, a Laravel application that consumes data contained within a WordPress instance over the JSON API is a HUUUUGE win. 

Allowing for even minimal installs of other applications to run along side of the WP installs we provision explicitly allows for this use case. We're not talking about provisioning mod_wsgi, Passenger, mod_python, etc., but merely other PHP-powered sites. 

*tl;dr version:* This is a simple step that allows us to move our core sites into a more regularized regime while simultaneously making HGV a much more attractive development platform.

File: provisioning/roles/wpsites/templates/etc/nginx/conf.d/wpsites.conf (hunk 8 above)
Comment (markkelnar) [inline provisioning/roles/wpsites/templates/etc/nginx/conf.d/wpsites.conf]: I'm more in favor of explicitly listing domains out in the yml config file.  We should discuss more on this and the format of the YML file next week.

Comment (zamoose) [inline provisioning/roles/wpsites/templates/etc/nginx/conf.d/wpsites.conf]: Let's definitely talk next week, but in brief, my justification for this is as follows:

One of the primary points to the initial HGV effort was to allow for testing of users' code in an environment that in many ways represents a close analogue to the Mercury systems. Allowing for A/B testing on traditional FPM vs. HHVM was an early goal and we've supported that through explicit core declarations. 

As we look to provide developers and users additional customization capabilities, it's my suggestion that we automate as much as possible. If we hard-code an assumption, that each site needs to be available in both FPM and HHVM versions, and then implement it as such, users will be unable to "forget" to provide the necessary framework for both versions to work. Instead, they provide a single site/URL declaration and we do the necessary legwork to provision both versions. (Well, in actuality, all 4 versions, with cache.x versions of each site + backend combo.)

Does that make sense?

Comment (zamoose) [inline provisioning/roles/wpsites/templates/etc/nginx/conf.d/wpsites.conf]: Addendum:
We also document this automation step so that users can rely upon those steps being accomplished by the provisioning framework.

**Comment by mrpritchett** [comment]
What's the status on this? Would love to be able to easily create multiple sites on the same box. Are there updated instructions anywhere else for current release?

**Comment by markkelnar** [comment]
@mrpritchett We are circling back on the feature of provisioning custom domains and multiple WP installs within the vagrant.  Hope to have an answer soon and for the next release.  Thanks for the interest in the feature.

**Comment by JasonTheAdams** [comment]
Just to chime in: This feature would be excellent and immediately make this vagrant a workflow standard for us!

**Comment by stephen-lin** [comment]
*NOTE: PR #133 is being broken into smaller, individual PRs. No further activity should happen here. This PR will remain open until the release of version 1.3.

**Comment by zamoose** [comment]
Closing this one out, since #153 and #154 effectively supplant its functionality.
